### PR TITLE
refactor: Remove FCNAME/FUNCNAME cruft

### PR DESCRIPTION
**Pull request by hzhou**
This is a revival of @akhillanger 's PR #3065. The PR is too good to see it get abandoned.

### CODE REVIEW

**Comment by hzhou** [comment]
Also reference Issue #2787

**Comment by hzhou** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by hzhou** [comment]
Two of the refactor commits are generated with two scripts. Following are the links to them:
  https://github.com/hzhou/mpich_study/blob/master/scripts/out/fix_FCNAME.pl
  https://github.com/hzhou/mpich_study/blob/master/scripts/out/fix_FCNAME_defs.pl

**Comment by hajimefu** [comment]
While I'm in favor of this direction of change in general, I'd prefer this to get merged together with the CH4 refactoring.

File: src/mpid/ch3/channels/nemesis/netmod/ofi/ofi_impl.h [hunk 1]
```diff
@@ -127,8 +127,8 @@ typedef struct {
 /* ******************************** */
 /* Logging and function macros      */
 /* ******************************** */
-#undef FUNCNAME
-#define FUNCNAME nothing
+
+/* FIXME: ref notes in src/mpid/ch4/shm/posix/posix_impl.h */
 #define BEGIN_FUNC(FUNCNAME)                    \
   MPIR_FUNC_VERBOSE_STATE_DECL(FUNCNAME);                   \
   MPIR_FUNC_VERBOSE_ENTER(FUNCNAME);
```
Comment (wesbland) [inline src/mpid/ch3/channels/nemesis/netmod/ofi/ofi_impl.h]: This macro should just go away.

File: src/mpid/ch4/shm/posix/posix_impl.h [hunk 2]
```diff
@@ -18,8 +18,12 @@
...
-#define FUNCNAME nothing
+/* FIXME: This should not work ...
+ * MPIR_FUNC_VERBOSE_STATE_DECL(a) etc uses #a to turn a into a string
+ * This macro is uses e.g. BEGIN_FUNC(__func__), but __func__ is a string
+ * So #a (as I quickly tested) simply become "__func__".
+ * Reviewers: please verify ...
+ */
 #define BEGIN_FUNC(FUNCNAME)                    \
     MPIR_FUNC_VERBOSE_STATE_DECL(FUNCNAME);     \
     MPIR_FUNC_VERBOSE_ENTER(FUNCNAME);
```
Comment (wesbland) [inline src/mpid/ch4/shm/posix/posix_impl.h]: Same thing here. Dump this macro.

Comment (hzhou) [reply src/mpid/ch4/shm/posix/posix_impl.h]: I agree on dumping these two macros. Shall we just revert the usage of `BEGIN_FUNC` and `END_FUNC` back to `MPIR_FUNC_VERBOSE_{STATE_DECL,ENTER,EXIT}`?

**Comment by hzhou** [comment]
> Also, do we have to fill in the actual function name in the `MPIR_FUNC_VERBOSE_STATE_DECL` macros (and the other similar ones)? Can we replace those with `__func__` too? Or even better, remove the name entirely and let them detect it from `__func__`?

The current `MPIR_FUNC_VERBOSE_...` set of macros use `#a` to turn the token into string. I don't think will work if `a` is a string such as `__func__` already -- my quick test seems to say it won't work.

I like the idea of using `__func__` always for these macros. They are debugging messages anyway, don't need to be perfect. But traditionally we have been coding up custom names with prefixes and such, so some people may have different preferences. Should we do it in a separate PR?

File: src/mpid/ch4/shm/posix/posix_impl.h (hunk 2 above)
Comment (wesbland) [reply src/mpid/ch4/shm/posix/posix_impl.h]: Yep.

**Comment by wesbland** [comment]
> The current `MPIR_FUNC_VERBOSE_...` set of macros use `#a` to turn the token into string. I don't think will work if `a` is a string such as `__func__` already -- my quick test seems to say it won't work.

Can we put the `__func__` inside the macros? Basically the macro would just be:

    #define MPIR_FUNC_ENTER() MPL_trvalid("Entering " __func__)

I don't know if that actually unrolls properly or not though.

> I like the idea of using `__func__` always for these macros. They are debugging messages anyway, don't need to be perfect. But traditionally we have been coding up custom names with prefixes and such, so some people may have different preferences. Should we do it in a separate PR?

I'd rather do all of this at once. Anyway, as @hajimefu said, my preference would be to do this at the same time as the other refactoring patches so all of us downstreams can do our rebases / merges at once.

**Comment by hzhou** [comment]
> I'd rather do all of this at once. Anyway, as @hajimefu said, my preference would be to do this at the same time as the other refactoring patches so all of us downstreams can do our rebases / merges at once.

We could schedule all refactor PRs to be committed at one batch. I prefer them to be reviewed early and then we can label it `READY_TO_COMMIT` and wait for the batch. 

I'll add the `MPIR_FUNC_` changes later, and later still, I would like to submit a PR for the default `if(mpi_errno)MPIR_ERR_POP(mpi_errrno)` with `MPIR_CALL`. I think doing separate PRs helps the review. Since most of them can be applied with script, it is not a big deal if I have to do a rebase at that batch commit time.

**Comment by hzhou** [comment]
I added a column -- "Batch Refactoring" -- to project "General Code Improvements" https://github.com/pmodels/mpich/projects/10

**Comment by hzhou** [comment]
> Can we put the __func__ inside the macros? Basically the macro would just be:
>
> ` #define MPIR_FUNC_ENTER() MPL_trvalid("Entering " __func__) `
>
@wesbland @pavanbalaji 
Turns out we can't. `__func__` is defined in C99 as if it is a string variable, and the preprocessor just sees it as `__func__` (not the string).

**Comment by pavanbalaji** [comment]
We can check for it in configure.

**Comment by wesbland** [comment]
Actually, when you look at the output in the debug logs, it already includes the file and line number:

    0   -1  103e565c0[51061]    1   0.000212    src/mpi/init/initthread.c   753 Entering MPID_STATE_MPI_INIT_THREAD

Is there any reason, we could just add a new debug macro that grabs that too? In `mpl_dbg.h`, `MPL_DBG_MSG` (which is what `MPIR_FUNC_ENTER` becomes) is defined as:

```
 #define MPL_DBG_MSG(_class, _level, _string)                            \
     {                                                                   \
         if ((_class & MPL_dbg_active_classes) && MPL_DBG_##_level <= MPL_dbg_max_level) { \
             MPL_dbg_outevent(__FILE__, __LINE__, _class, 0, "%s", _string); \
         }                                                               \
     }
```

Let's just create a new macro like this:

```
 #define MPL_DBG_MSG_FUNC_ENTER_EXIT(_class, _level, _string)                            \
     {                                                                   \
         if ((_class & MPL_dbg_active_classes) && MPL_DBG_##_level <= MPL_dbg_max_level) { \
             MPL_dbg_outevent(__FILE__, __LINE__, _class, 0, "%s %s", _string, __func__); \
         }                                                               \
     }
```

We can just drop the `MPIR_FUNC_VERBOSE_STATE_DECL` macros as they're always defined to be no ops.

Then we just need to leave the `MPIR_FUNC_VERBOSE_ENTER` and `MPIR_FUNC_VERBOSE_EXIT` macros without any arguments. Less boilerplate == better. 😄

**Comment by hzhou** [comment]
`_STATE_DECL_` is not always defined as NOOP, for example: `src/include/rlog_macros.h`. 

New macros for `MPL_DBG_MSG_` seems OK. In fact, since these macros are only used with `MPIR_FUNC_...`, which no longer fits int `MPL_` namespace,  we might as well roll them into `src/include/mpir_func.h`.

**Comment by wesbland** [comment]
> `_STATE_DECL_` is not always defined as NOOP, for example: `src/include/rlog_macros.h`.

Hmm...my ctags let me down. Fair enough.

Still...removing the need for the arguments is still a good win.

**Comment by hzhou** [comment]
@wesbland Take a look at PR #3538, what do you think?

**Comment by hzhou** [comment]
> I'm much more inclined to just leave in the empty macros as suggested on #3534 since that removes pretty much all of my concerns.

@wesbland I believe you were referring to this PR.  I believe by "empty macros" you meant that macros `MPIR_FUNC_...` not taking a parameter. I tried this morning to do just that and realized that we can't really do that without a whole bunch of complications.  If we only using these macros to leave a debug log, then you are right that this can be easily done by redefining the macros in `mpir_func.h`. However, the current architecture also allows configuring `MEMARENA` or `rlog` or potentially arbitrary future additions. `MEMARENA` functions and (especially) `rlog` facility are designed to deal with more general usage not just function entering and exiting -- regardless of what they have been used for. So in order to roll `__func__` into these facilities, we need narrow down their designs or fork a special version of these facilities. It is unclear how often we use them practically so the effort seems ill-warranted.

It will be an easier (trivial) issue if we can say that debugging message is all we wanted. But that is a discussion beyond refactoring.

So the `BEGIN_FUN` and `END_FUNC` refactoring is broken. I need to revert/fix it to the use explicit function ids.

**Comment by wesbland** [comment]
How about instead of having empty macros, we just put `__func__` in the macros? Then we have the string in the right place, right?

**Comment by hzhou** [comment]
> How about instead of having empty macros, we just put __func__ in the macros? Then we have the string in the right place, right?

Unfortunately that won't work. `__func__` is only a string at compile time, not preprocessing time. You end up with "Enrering __func__".

**Comment by raffenet** [comment]
Shall we revive this effort now that a good portion of the major ch4 refactoring has been merged?

**Comment by hzhou** [comment]
> 
> 
> Shall we revive this effort now that a good portion of the major ch4 refactoring has been merged?

Sounds good. I'll revive it. 

The `maint/extractstates.in` uses `#define FUNCNAME ...` line to parse function signatures. This script is used for `rlog` feature (as far as I know). The `rlog` is quite broken as now, shall we make the decision to completely remove the `rlog` support from the build system?

**Comment by raffenet** [comment]
I would vote remove `rlog` support. It has not worked for some time and there has not been strong interest in reviving it.

**Comment by hzhou** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by hzhou** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by hzhou** [comment]
test:jenkins/ch3/most 
test:jenkins/ch4/most

**Comment by hzhou** [comment]
`__func__` is 2 characters longer than `FCNAME`, and that caused a bunch of whitespace checker errors ... come on :(

I think that long line rule should be removed. 1 the 100 char limit is arbitrary. 2. the automatic reformat by gnu indent does not group parameters by semantics, often does not enhance readability. 3 We have quite a few long functions names -- forcing single or few parameters/arguments on each new line -- definitely makes the code grosser IMO. 4 I believe we are losing more productivity on this one rule than its benefit (if any).

**Comment by wesbland** [comment]
I'm ok with removing the rule, but what do we do about all the code already
formatted this way?

>

**Comment by hzhou** [comment]
> I'm ok with removing the rule, but what do we do about all the code already formatted this way?

I don't think  `gnu indent` will do/complain anything with manually split lines that do not exceed the limit. I routinely manually break up lines (according to semantic grouping) and seems fine. So the answer to your question could be "nothing".

**Comment by wesbland** [comment]
👍

>

**Comment by pavanbalaji** [comment]
Please verify and confirm on this PR that this doesn’t actually go and reformat all of the code.  That’d be silly.

**Comment by yfguo** [comment]
If we remove the character limit, the `code-cleanup.sh` will concatenate the short lines into a long one. So unless we want to do another code cleanup, I think we would have to just live with it.

I do not know how did we set the 100 character limit. But I think it is better than no limit. The problem with no character limit is any other tools involved in code dev/viewing must have line wrap support. That include debuggers, diff tools and Github. Among them, Github does not have line wrap (not when the page is displayed at full width). The debuggers are likely not having it or not enabling it by default. The line wrap does not respect the indent. The long line will be displayed in an ugly way.

**Comment by pavanbalaji** [comment]
We are not going to do another overhaul code cleanup till we come up with ch5.

**Comment by hzhou** [comment]
> If we remove the character limit, the `code-cleanup.sh` will concatenate the short lines into a long one. 

I will be very surprised if `code-cleanup.sh` do that. I'll verify it later.

> I do not know how did we set the 100 character limit. But I think it is better than no limit. The problem with no character limit is any other tools involved in code dev/viewing must have line wrap support. That include debuggers, diff tools and Github. Among them, Github does not have line wrap (not when the page is displayed at full width). The debuggers are likely not having it or not enabling it by default. The line wrap does not respect the indent. The long line will be displayed in an ugly way.

I would believe tools will work more easily with unwrapped lines than wrapped ones. I always get annoyed that gdb stepping through partial lines. Github could be a great reason that even though we don't set hard limit, we still will keep the lines in reasonable length -- because reviewers will demand it.

**Comment by hzhou** [comment]
> We are not going to do another overhaul code cleanup till we come up with ch5.

The premises is we don't need do anything. We are removing a restriction, not adding one. And another premises is we will still keep our code lines in reasonable length even without the hard limit.

**Comment by pavanbalaji** [comment]
That’s just a hypothesis without a confirmation.  Let’s confirm that before arguing about it.

**Comment by hzhou** [comment]
To my surprise, `GNU indent` will join my deliberately-split beautiful lines! I searched hi and lo, and there isn't an option to override it. I guess I have to bow to the tool and work with `--line-length100`. 

PS: this is silly. To me, not to touch my line breaks seems basic and trivial; but I guess the way `GNU indent` is written, it is very hard to do. 

Close PR #3732.

**Comment by hzhou** [comment]
> 
> 
> That’s just a hypothesis without a confirmation. Let’s confirm that before arguing about it.

I confirm my hypotheses were wrong.

**Comment by hzhou** [comment]
I amended the commits to fix the `whitespace-checker`. The last commit now includes the whitespace that we missed last time -- most of them from my `gentran` renaming patch. If we split the whitespace change from the main change into separate commits, then the `whitespace-checker` won't pass; and if we ignore, then we risk leaking in the whitespace errors. I think the lesser evil is to insist each commit from now on always whitespace clean before merge,

**Comment by hzhou** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by wesbland** [comment]
If you want to give credit to @akhillanger, the best way to do that is with the co-author tag since that's what tools like GitHub use:

    Co-authored-by: "Akhil Langer <akhil.langer@intel.com>"

**Comment by hzhou** [comment]
> If you want to give credit to @akhillanger, the best way to do that is with the co-author tag since that's what tools like GitHub use:
> 
> ```
> Co-authored-by: "Akhil Langer <akhil.langer@intel.com>"
> ```

Will do.

**Comment by hzhou** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by hzhou** [comment]
@raffenet Do you want to give it a final review?

**Pull request merged**